prefect_kubernetes/worker.py

```diff
@@ -0,0 +1,671 @@
+        Parse a Kubernetes infrastructure PID into its component parts.
+
+        Returns a cluster UID, namespace, and job name.
+        """
+        cluster_uid, namespace, job_name = infrastructure_pid.split(":", 2)
+        return cluster_uid, namespace, job_name
+
+    @contextmanager
+    def _get_core_client(
+        self, client: "ApiClient"
+    ) -> Generator["CoreV1Api", None, None]:
+        """
+        Context manager for retrieving a Kubernetes core client.
+        """
+        try:
+            yield kubernetes.client.CoreV1Api(api_client=client)
+        finally:
+            client.rest_client.pool_manager.clear()
+
+    def _get_cluster_uid(self, client: "ApiClient") -> str:
+        """
+        Gets a unique id for the current cluster being used.
+
+        There is no real unique identifier for a cluster. However, the `kube-system`
+        namespace is immutable and has a persistence UID that we use instead.
+
+        PREFECT_KUBERNETES_CLUSTER_UID can be set in cases where the `kube-system`
+        namespace cannot be read e.g. when a cluster role cannot be created. If set,
+        this variable will be used and we will not attempt to read the `kube-system`
+        namespace.
+
+        See https://github.com/kubernetes/kubernetes/issues/44954
+        """
+        # Default to an environment variable
+        env_cluster_uid = os.environ.get("PREFECT_KUBERNETES_CLUSTER_UID")
+        if env_cluster_uid:
+            return env_cluster_uid
+
+        # Read the UID from the cluster namespace
+        with self._get_core_client(client) as core_client:
+            namespace = core_client.read_namespace("kube-system")
+        cluster_uid = namespace.metadata.uid
+
+        return cluster_uid
+
+    def _watch_job(
+        self,
+        job_name: str,
+        configuration: KubernetesWorkerJobConfiguration,
+        client: "ApiClient",
+    ) -> int:
+        """
+        Watch a job.
+
+        Return the final status code of the first container.
+        """
+        self._logger.debug(f"Job {job_name!r}: Monitoring job...")
+
+        job = self._get_job(job_name, configuration, client)
+        if not job:
+            return -1
+
+        pod = self._get_job_pod(job_name, configuration, client)
+        if not pod:
+            return -1
+
+        # Calculate the deadline before streaming output
+        deadline = (
+            (time.monotonic() + configuration.job_watch_timeout_seconds)
+            if configuration.job_watch_timeout_seconds is not None
+            else None
+        )
+
+        if configuration.stream_output:
+            with self._get_core_client(client) as core_client:
+                logs = core_client.read_namespaced_pod_log(
+                    pod.metadata.name,
+                    configuration.namespace,
+                    follow=True,
+                    _preload_content=False,
+                    container="prefect-job",
+                )
+                try:
```

Comment:
do we need retries here? the log stream connection can break if there are connectivity issues between the agent and Kubernetes cluster